### PR TITLE
Fix incorrect usage of `asyncio` in ledger connection

### LINKED ISSUE #1173: asyncio.wait() copies the task container that is passed to it
### Subject of the issue
Contract & ledger operations can be blocked if a previously sent request to the `ledger` connection is blocked. The issue stems from a wrong usage of the `asyncio.wait()` method. 
We know for sure the `ledger` connection is affected, but any other component that uses `asyncio.wait()` in the way described below is affected by it. 

#### What asyncio.wait() does
Let `tasks` be an iterable of tasks (or futures), if we want to wait for ANY of the tasks to be completed, we can do so with `asycnio.wait()` in the following way:

 ```python
 done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_COMPLETED)
 ```
The line above will block until at least one of the tasks is done. The completed tasks will be assigned to `done` and those that are still running will be assigned to `pending`. 
Note that its irrelevant as to what the underlying tasks are running and through which mechanism they are running, this includes tasks running on worker threads.

In the [ledger connection](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py) we can see this being used in the `receive` method:
```python
    async def receive(self, *args: Any, **kwargs: Any) -> Optional["Envelope"]:
        """
        Receive an envelope. Blocking.
        :param args: the arguments
        :param kwargs: the keyword arguments
        :return: the envelope received, or None.
        """
        # if there are done tasks, return the result
        if len(self.done_tasks) > 0:  # pragma: nocover
            done_task = self.done_tasks.pop()
            return self._handle_done_task(done_task)

        if len(self.receiving_tasks) == 0:  # pragma: nocover
            self.event_new_receiving_task.clear()
            await self.event_new_receiving_task.wait()

        # wait for completion of at least one receiving task
        done, _ = await asyncio.wait(
            self.receiving_tasks, return_when=asyncio.FIRST_COMPLETED
        )

        # pick one done task
        done_task = done.pop()

        # update done tasks
        self.done_tasks.extend([*done])

        return self._handle_done_task(done_task)
```
 
In the `receiving_tasks` have a list of tasks that are being executed in the background, we await for at least one of them to be completed so that we send a response back:
```python
        done, _ = await asyncio.wait(
            self.receiving_tasks, return_when=asyncio.FIRST_COMPLETED
        )
```
#### Where's the problem?
`asyncio.wait()` [copies the iterable](https://github.com/python/cpython/blob/v3.10.4/Lib/asyncio/tasks.py#L374) that is being passed to it.
Consider the following:
```python
tasks = []

async def receive():
  done, _ = await asyncio.wait(
     tasks, return_when=asyncio.FIRST_COMPLETED
  )

async def send(task):
  tasks.append(task)

send(task_a)
receive_task = asyncio.create_task(receive())
send(task_b)
await receive_task
```
In the current scenario, even though we have two tasks in `tasks`, we are only waiting for `task_a` to be finished, because `asyncio.wait()` [copies the list](https://github.com/python/cpython/blob/v3.10.4/Lib/asyncio/tasks.py#L374) that is being passed to it.

In the ledger connection we have the same setup as the example above.
We have a list of [receiving_tasks](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L175) that represents requests being processed. And in the send method we [append to this list](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L125).

Now consider the following flow of events:
1. Assume a fresh instance of the ledger connection.
2. Assume the multiplexer has created a task `task_receive` by calling the [receive method](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L172) of the ledger connection. Since the ledger connection is fresh, `task_receive` will be [awaiting a request to be received](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L172).
3. A request is sent to the ledger connection. Ex: a `transaction_receipt` request. A task `task_a` is [created for this request](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L125). 
4. The `event_new_receiving_task` event [is triggered](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L127), and `task_receive` will now be [awaiting for `task_a`  to complete](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L175).
5. Assume `task_a` is blocked. This can happen for many reasons, a bad node, crypto opts taking longer than usual etc.
6. Assume a contract request is sent to the ledger connection. A task `task_b` is created for this request, and [is added to the receiveing tasks](https://github.com/valory-xyz/open-autonomy/blob/v0.1.6/packages/valory/connections/ledger/connection.py#L125).
7. Assume `task_b` finishes 3 seconds after its creation. Because we are only waiting for `task_a` to finish, `task_b` will wait until `task_a` is finished.

The following snippet reproduces this:
```python
import asyncio


async def blocking_task():
    print("started blocking task")
    await asyncio.sleep(10000)
    print("finished blocking task")


async def normal_task():
    print("started normal task")
    await asyncio.sleep(1)
    print("finished normal task")


tasks = []


async def receive():
    done, _ = await asyncio.wait(
        tasks, return_when=asyncio.FIRST_COMPLETED
    )


def send(task):
    tasks.append(task)


async def main():
    # a blocking task is sent to the con
    send(asyncio.create_task(blocking_task()))

    # the multiplexer awaits for any task to be finished
    receive_task = asyncio.create_task(receive())
    receive_task.add_done_callback(lambda _: print("done"))

    # 3 seconds later a normal task is received
    await asyncio.sleep(3)
    send(asyncio.create_task(normal_task()))

    # the normal task will finish 1 second later but the receive task won't be completed,
    # i.e. "done" won't be printed
    await asyncio.sleep(1000)


asyncio.get_event_loop().run_until_complete(main())
```

### PR DESCRIPTION
**Pull request by Adamantios**
## Proposed changes

Fixes the ledger connection issue, caused by wrong usage of the `asyncio.wait()` method.
Replaces #1195.

## Fixes

Closes #1173.

## Types of changes

What types of changes does your code introduce? (A **breaking change** is a fix or feature that would cause existing functionality and APIs to not work as expected.)
_Put an `x` in the box that applies_

- [x] Non-breaking fix (non-breaking change which fixes an issue)
- [ ] Breaking fix (breaking change which fixes an issue)
- [ ] Non-breaking feature (non-breaking change which adds functionality)
- [ ] Breaking feature (breaking change which adds functionality)
- [ ] Refactor (non-breaking change which changes implementation)
- [ ] Messy (mixture of the above - requires explanation!)

## Checklist

_Put an `x` in the boxes that apply._

- [x] I have read the [CONTRIBUTING](../blob/main/CONTRIBUTING.md) doc
- [x] I am making a pull request against the `main` branch (left side). Also you should start your branch off our `main`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have locally run services that could be impacted and they do not present failures derived from my changes

## Further comments

n/a

### CODE REVIEW

File: packages/valory/connections/ledger/connection.py [hunk 1]
```diff
@@ -67,9 +63,9 @@ def __init__(self, **kwargs: Any):
         )
 
     @property
-    def event_new_receiving_task(self) -> asyncio.Event:
-        """Get the event to notify the 'receive' method of new receiving tasks."""
-        return cast(asyncio.Event, self._event_new_receiving_task)
+    def response_envelopes(self) -> asyncio.Queue:
+        """Get the response envelopes. Only intended to be accessed when connected."""
+        return cast(asyncio.Queue, self._response_envelopes)
```
Comment (DavidMinarsch) [inline packages/valory/connections/ledger/connection.py:68]:
```suggestion
        if self._response_envelopes is None:
            raise ValueError("asyncio.Queue not set")
        return self._response_envelopes
```

Comment (Adamantios) [reply packages/valory/connections/ledger/connection.py]: Done in d2ac6547dd2d77e3e009ce936ecbe8d4a027f5a5.

**Review by DavidMinarsch — APPROVED** [review]

**Review by 0xArdi — APPROVED** [review]

**Comment by Karrenbelt** [comment]
```
FAILED tests/test_skills/test_counter_client/test_behaviours.py::TestIncrementerBehaviour::test_run
FAILED tests/test_autonomy/test_cli/test_scaffold_fsm.py::TestScaffoldFSM::test_run
FAILED tests/test_skills/test_counter_client/test_behaviours.py::TestHttpHandler::test_handle
FAILED tests/test_skills/test_counter_client/test_behaviours.py::TestMonitorBehaviour::test_run
```
please [add here](https://docs.google.com/document/d/1f7DrPdTXaLWJ1icTZ1Mt2S3z3ZdzCodrxAG6d1tI3o0) so we can squash them when time is due

File: packages/valory/connections/ledger/connection.py [hunk 2]
```diff
@@ -162,37 +161,15 @@ async def receive(self, *args: Any, **kwargs: Any) -> Optional["Envelope"]:
...
-            await self.event_new_receiving_task.wait()
-
-        # wait for completion of at least one receiving task
-        done, _ = await asyncio.wait(
-            self.receiving_tasks, return_when=asyncio.FIRST_COMPLETED
-        )
-
-        # pick one done task
-        done_task = done.pop()
+        return await self.response_envelopes.get()
```
Comment (Karrenbelt) [inline packages/valory/connections/ledger/connection.py:164]: would it be better to use a timeout here?
 `await asyncio.wait_for(self.agent.outbox.get(), timeout=1)`

File: packages/valory/connections/ledger/connection.py [hunk 3]
```diff
@@ -203,4 +180,6 @@ def _handle_done_task(self, task: asyncio.Future) -> Optional[Envelope]:
                 message=response_message,
                 context=request.context,
             )
-        return response_envelope
+
+        # not handling `asyncio.QueueFull` exception, because the maxsize we defined for the Queue is infinite
+        self.response_envelopes.put_nowait(response_envelope)
```
Comment (Karrenbelt) [inline packages/valory/connections/ledger/connection.py:185]: are you going to address the full queue issue as well? Leaving it at infinite size seems like a bad idea, not?

File: packages/valory/connections/ledger/connection.py [hunk 4]
```diff
@@ -106,12 +106,12 @@ async def disconnect(self) -> None:
 
         self.state = ConnectionStates.disconnecting
 
-        for task in self.receiving_tasks:
+        for task in self.task_to_request.keys():
```
Comment (Karrenbelt) [inline packages/valory/connections/ledger/connection.py:109]: `.keys()` not needed

File: packages/valory/connections/ledger/connection.py (hunk 2 above)
Comment (DavidMinarsch) [reply packages/valory/connections/ledger/connection.py:164]: The whole point is that it doesn't time out.

File: packages/valory/connections/ledger/connection.py (hunk 3 above)
Comment (DavidMinarsch) [reply packages/valory/connections/ledger/connection.py:185]: That would introduce an arbitrary magic value defining the queue size. It's not a bad question, but the answer is more involved as there are other areas of the framework where queues can build up.

**Pull request merged**